pkg/virt-handler/heartbeat/heartbeat.go

```diff
@@ -71,6 +71,24 @@ func (h *HeartBeat) heartBeat(heartBeatInterval time.Duration, stopCh chan struc
 	// 1 minute with a 1.2 jitter + the time it takes for the heartbeat function to run (sliding == true).
 	// So the amount of time between heartbeats randomly varies between 1min and 2min12sec + the heartbeat function execution time.
 	wait.JitterUntil(h.do, heartBeatInterval, 1.2, true, stopCh)
+	now, err := json.Marshal(metav1.Now())
```

Comment:
I've modified the code a bit to put the logic I added into a separate function and added a done channel. Is this what you were suggesting in order to make this stable?

Reply:
Sorry for being a little bit cryptic. We have this place: https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L1400. There we start the heartbeat in go-routine and it just runs until the stop channel gets closed.

The controller-shutdown-procedure will not wait for it. It just also waits on the same stop channel: https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm.go#L1407

There then basically happens a shutdown race:

 * heartbeat tries to stop and do a last http reqeust
 * controller shuts down
 
 Right now: If the controller shuts down first, then the main function just exists, killing all remaining go routines right away.
 
 If the rest call from the heartbeat then happens or not is basically a matter of luck. If we would start the heartbeat like this:
 
 
 ```
 
heartBeatDone := make(chan struct{})
 go func() {
 	c.heartBeat.Run(c.heartBeatInterval, stopCh)
 	close(heartBeatDone)
}()

<- heartBeatDone
<- stopCh
 ```
 
 we could the heartbeat guarantee a single http request to mark itself as unschedulable. If not, we have to ensure that the node controller sets the schedulable condition to false once we reach the heartbeat timeout (I think it happens, but since it takes a little bit you may not have observed it).

Reply:
Okay, I've added the done channel to the heartBeat.run call in vm.go, could you let me know what you think?